overlord/configstate/configcore/netplan.go

```diff
@@ -0,0 +1,129 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package configcore
+
+import (
+	"fmt"
+	"strings"
+
+	// TODO: Move to yaml.v3 everywhere, there is PR#10696 that starts
+	//       this. However it is not trivial yaml.v2 accepts duplicated
+	//       keys in maps and v3 does not. There might be snaps in the
+	//       wild that we could break by going to v3.
+	//
+	// Move this part of the code to yaml.v3 because without it we run
+	// into incompatibilites of maps between json and yaml:
+	// "json: unsupported type: map[interface {}]interface {}" because
+	// because yaml.v2 unmarshalls by default to "map[string]interface{}"
+	// v3 fixes this, see https://github.com/go-yaml/yaml/pull/385#issuecomment-475588596
+	"gopkg.in/yaml.v3"
+
+	"github.com/godbus/dbus"
+	"github.com/godbus/dbus/introspect"
+
+	"github.com/snapcore/snapd/dbusutil"
+	"github.com/snapcore/snapd/overlord/configstate/config"
+	"github.com/snapcore/snapd/release"
+)
+
+func init() {
+	// add supported configuration of this module
+	supportedConfigurations["core.system.network.netplan"] = true
+	// and register as exteranl config
+	config.RegisterExternalConfig("core", "system.network.netplan", getNetplanFromSystem)
+}
+
+func validateNetplanSettings(tr config.Conf) error {
+	// XXX: validate somehow once we support writing?
+	return nil
+}
+
+func handleNetplanConfiguration(tr config.Conf, opts *fsOnlyContext) error {
+	for _, chg := range tr.Changes() {
+		if strings.HasPrefix(chg, "core.system.network.netplan.") {
+			return fmt.Errorf("cannot set netplan config yet")
+		}
+	}
+
+	return nil
+}
+
+func hasDBusMethodOnInterface(node *introspect.Node, ifName, methodName string) bool {
+	for _, iff := range node.Interfaces {
+		if iff.Name == ifName {
+			for _, mth := range iff.Methods {
+				if mth.Name == methodName {
+					return true
+				}
+			}
+		}
+	}
+	return false
+}
+
+func getNetplanFromSystem(key string) (result interface{}, err error) {
+	if release.OnClassic {
+		return nil, nil
+	}
+
+	conn, err := dbusutil.SystemBus()
+	if err != nil {
+		return nil, err
+	}
+	// godbus uses a global systemBus object internally so we *must*
+	// not close the connection.
+
+	var netplanConfigSnapshotBusAddr string
+	netplan := conn.Object("io.netplan.Netplan", "/io/netplan/Netplan")
+
+	// introspect
+	node, err := introspect.Call(netplan)
+	if derr, ok := err.(dbus.Error); ok {
+		// ignore if there is no dbus service for netplan
+		if derr.Name == "org.freedesktop.DBus.Error.ServiceUnknown" {
+			return nil, nil
+		}
+	}
+	// XXX: should we log and return nil here to avoid breaking the
+	// conf system if something is wrong with dbus on the machine?
+	if err != nil {
+		return nil, err
+	}
+	if !hasDBusMethodOnInterface(node, "io.netplan.Netplan", "Config") {
```

Comment:
This probably deserves a comment: I would think that if the interface and method name are missing, this could be treated as an error (which you'd receive as a result of the `netplan.Call("io.netplan.Netplan.Config", 0)` below.

If we really want to treat the absence of this method specially, we should somehow make this more explicit.

Reply:
Ah, I understood it now by looking at the tests. Still, I vote for adding a comment here, along these lines: "If the service implements only the netplan v1 API, which was missing the Config method, pretend that netplan is not there at all".

By the way, it's probably more efficient if we skip the introspection, and instead handle this case by looking at the error received by the `Config` call below. So, in the happy case, we'd just make one D-Bus call instead of two.